lib/graph_gateway.go

```diff
@@ -315,14 +318,46 @@ func (api *GraphGateway) loadRequestIntoSharedBlockstoreAndBlocksGateway(ctx con
 			metrics.contextAlreadyCancelledMetric.Inc()
 		}
 
-		err := api.fetcher.Fetch(ctx, path, func(resource string, reader io.Reader) error {
+		cctx, cncl := context.WithCancel(ctx)
+		defer cncl()
+		err := api.fetcher.Fetch(cctx, path, func(resource string, reader io.Reader) error {
 			cr, err := car.NewCarReader(reader)
 			if err != nil {
 				return err
 			}
+
+			blkCh := make(chan blocks.Block, 1)
+			go func() {
+				defer close(blkCh)
+				for {
+					blk, rdErr := cr.Next()
+					if rdErr != nil {
+						err = rdErr
+					}
+					blkCh <- blk
+				}
+			}()
+
+			t := time.NewTimer(GetBlockTimeout)
```

Comment:
is your worry that we'd end up failing with `ErrNoProgress` when we should fail with the error surfaced from the timeout of `cr.Next()`?

we can make the initial timer 2x block timeout so that we get the underlying error in that case

Reply:
My concern is that if we get sent back the CAR header but not the first block that we'll wait 30 minutes instead of 1. Am I misreading how this works?

Reply:
Now it's just the header, which is fine that can have it's own timeout handled by the Fetch implementation. IIRC before the first block wasn't controlled by a block request timeout either, but it is now (with a 2x GetBlockTimeout).